com.dynamo.cr/com.dynamo.cr.bob/src/com/dynamo/bob/util/ComponentsCounter.java

```diff
@@ -0,0 +1,323 @@
+import com.dynamo.bob.pipeline.ProtoUtil;
+import com.dynamo.bob.pipeline.BuilderUtil;
+import com.dynamo.gameobject.proto.GameObject.ComponenTypeDesc;
+import com.dynamo.gameobject.proto.GameObject.CollectionDesc;
+import com.dynamo.gameobject.proto.GameObject.ComponentDesc;
+import com.dynamo.gameobject.proto.GameObject.EmbeddedComponentDesc;
+import com.dynamo.gameobject.proto.GameObject.PrototypeDesc;
+import com.dynamo.gamesys.proto.GameSystem.CollectionFactoryDesc;
+import com.dynamo.gamesys.proto.GameSystem.FactoryDesc;
+
+/**
+ * Class helps to count components in GOs and Collections
+ */
+public class ComponentsCounter {
+
+    public static final String EXT_GO = ".compcount_go";
+    public static final String EXT_COL = ".compcount_col";
+    public static final Integer DYNAMIC_VALUE = 0xFFFFFFFF;
+
+    /**
+     * Class represents a storage for counted components which can be written on disk
+     */
+    public static class Storage implements Serializable {
+
+        private static final long serialVersionUID = 1L;
+
+        private Map<String, Integer> components = new HashMap<>();
+        private Boolean isInDynamicFactory = false;
+
+        public Map<String, Integer> get() {
+            return components;
+        }
+
+        public void makeDynamic() {
+            isInDynamicFactory = true;
+        }
+
+        public Boolean isDynamic() {
+            return isInDynamicFactory;
+        }
+
+        public void add(String componentName, Integer count) {
+            if (componentName.charAt(0) == '.') {
+                componentName = componentName.substring(1);
+            }
+            Integer currentValue = components.getOrDefault(componentName, 0);
+            if (count == DYNAMIC_VALUE) {
+                components.put(componentName, DYNAMIC_VALUE);
+            } 
+            else if (currentValue != DYNAMIC_VALUE) {
+                components.put(componentName, currentValue + count);
+            }
+        }
+
+        public void add(String componentName) {
+            add(componentName, 1);
+        }
+
+        public void add(Storage storage) {
+            Map<String, Integer> comps = storage.get();
+            for (Map.Entry<String,Integer> entry : comps.entrySet()) {
+                add(entry.getKey(), entry.getValue());
+            }
+        }
+
+        public void add(Storage storage, Integer count) {
+            Map<String, Integer> comps = storage.get();
+            for (Map.Entry<String,Integer> entry : comps.entrySet()) {
+                add(entry.getKey(), count == DYNAMIC_VALUE ? DYNAMIC_VALUE : entry.getValue() * count);
+            }
+        }
+
+        public byte[] toByteArray() throws IOException {
+            ByteArrayOutputStream bos = new ByteArrayOutputStream(128 * 16);
+            ObjectOutputStream os = new ObjectOutputStream(bos);
+            os.writeObject(this);
+            os.close();
+            return bos.toByteArray();
+        }
```

Comment:
For debuggability, would it be possible to store them as text? (I assume there's some Java class that helps with this)

Reply:
I thin it's easier to cover with extra logging (esp taking into account that we have collections which counts embeded objects themself and can't take date from embeded objects' `compcount_go`)